feat: Retry failed media uploads when re-establishing network connection

dcalhoun opened the pull request:
## Description

Improve the media upload experience in the post editor by automatically retrying
failed uploads when re-establishing network connectivity.

Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/6406.

## Testing Instructions

### When network connectivity is available
1. Launch the editor.
1. Add several media blocks, attaching media to the blocks.
1. Disconnect your device network connection.
1. Note the failed state of media uploads.
1. Connect your device network connection.
1. Verify the media uploads automatically restart and succeed.

### When network connectivity is unavailable
1. Launch the editor.
1. Add several media blocks, attaching media to the blocks.
1. Note the failed state of media uploads.
1. Connect your device network connection.
1. Verify the media uploads automatically restart and succeed.

## Regression Notes
1. Potential unintended areas of impact
    Editor media uploads, editor launch, writing flow.
2. What I did to test those areas of impact (or what existing automated tests I relied on)
    Insert and interact with various blocks using the editor.
3. What automated tests I added (or what prevented me from doing so)
    None, lack of familiarity with the testing stack.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

dcalhoun committed:
feat: Retry failed media uploads when re-establishing network connection

Improve the media upload UX by automatically retrying failed uploads.

dcalhoun committed:
docs: Add release note

dcalhoun commented on libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -1583,5 +1583,9 @@ public void onGutenbergDialogNegativeClicked(@NonNull String instanceTag) {
         getGutenbergContainerFragment().onConnectionStatusChange(isConnected);
+        if (hasFailedMediaUploads()) {
+            mEditorFragmentListener.onMediaRetryAll(mFailedMediaIds);
+            mFailedMediaIds.clear();
```
Not sure clearing this is necessary, it does not appear to be done elsewhere, e.g., [`showRetryMediaUploadDialog`](https://github.com/wordpress-mobile/WordPress-Android/blob/f4bc1f62fe6b0be7f42ae4bbfed60f6931e5982a/libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L917-L944). This clear was added in an attempt to fix erroneously displayed errors after saving or publishing a post, but it does not appear to work. 

<details><summary>Error screenshot</summary>

![Unexpected failed media upload error after closing editor](https://github.com/wordpress-mobile/WordPress-Android/assets/438664/d7b59e78-cbf2-4e8a-a0f4-724d2ea06fc0)

</details>

<details><summary>Steps to reproduce</summary>

1. Use the prototype build on this PR. 
1. Disable network connectivity.
2. Add an Image block and attach media. 
3. Enable network connectivity. 
4. Allow the automatic retry of media upload to finish successfully. 
5. Save and close the post. 
6. Note the failed media upload notification. 

</details>

It appears as though a lingering error state/message is in place, but I am unable to find where that is or how to clear it. I expected [`onMediaRetryAll`](https://github.com/wordpress-mobile/WordPress-Android/blob/f4bc1f62fe6b0be7f42ae4bbfed60f6931e5982a/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L3303-L3311) to clear it, but it doesn't appear to be doing so. 

👋🏻 @antonis. Randomly pinging you for Android expertise and help. 😄 Would you be willing to help me debug why I might see this error when the media attached to the post actually _successfully uploads_ with the new retry logic?

dcalhoun replied:
After testing the existing functionality of retrying failing media uploads via the dialog presented when tapping a failed image, it appears the lingering error message is a preexisting issue. Semi-related issues are reported in https://github.com/wordpress-mobile/WordPress-Android/issues/11039, https://github.com/wordpress-mobile/WordPress-Android/issues/15897, and https://github.com/wordpress-mobile/WordPress-Android/issues/13427. 

<details><summary>Steps to reproduce</summary>

1. Use the latest Play Store app version. 
1. Disable network connectivity.
2. Add an Image block and attach media. 
3. Enable network connectivity. 
4. Tap the Image block with a failed upload. 
5. Tap "Retry."
6. After the media finishes uploading successfully, save and close the post. 
7. Note the failed media upload notification. 

</details>

Therefore, I do not consider this issue blocking. It'd be nice to resolve, but could be done so in a separate PR later. 

@antonis please do not feel like it is a priority to help debug this. Thanks!

antonis replied:
> Not sure clearing this is necessary

Removing the clearing with c114528b47a56b3f9394acdf771b5c393a3447fb and the reasoning for it sounds good 👍 

> Therefore, I do not consider this issue blocking. It'd be nice to resolve, but could be done so in a separate PR later.

I was able to reproduce this issue and tried to figure out why this happens. My understanding is that there is a limitation on how [the `SnackbarSequencer`](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/java/org/wordpress/android/util/SnackbarSequencer.kt) works were this kind of internal notifications are not really canceled once issued. I agree that this can be investigated further and handled in a follow up PR 👍

dcalhoun replied:
Thank you for investigating and sharing your insights and perspective. Very helpful! 🙇🏻‍♂️

dcalhoun committed:
feat: Limit automatic upload retries to debug builds

We will enable the feature in production once all project work is
completed.

dcalhoun committed:
refactor: Remove unnecessary clearing of failed media IDs

This was originally added in an attempt to clear an unexpected "failed
media upload" notification displayed after closing the editor. However,
it did not resolve that issue, it must originate from some other
location. Additionally, it would appears successful media uploads
already manage removing media IDs from the `mFailedMediaIds` collection.

antonis commented on libs/editor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -1583,5 +1583,8 @@ public void onGutenbergDialogNegativeClicked(@NonNull String instanceTag) {
     @Override
     public void onConnectionStatusChange(boolean isConnected) {
         getGutenbergContainerFragment().onConnectionStatusChange(isConnected);
+        if (BuildConfig.DEBUG && hasFailedMediaUploads()) {
```
```suggestion
        if (BuildConfig.DEBUG && isConnected && hasFailedMediaUploads()) {
```
I would suggest adding an extra `isConnected` condition is it is available in this scope though it might not be necessary.

dcalhoun replied:
Ooh! A very good suggestion. Thank you.

antonis reviewed (approved):
Thank you for fixing this @dcalhoun 🙇 
I tested the implementation in debug mode (47df15049648509ec3346ad10e2867c793735ead) and the app performed as expected. The code also looks good and I only added a minor suggestion 🎉 
Overall I think this is an improvement to the current user experience when uploads fail 🏅

dcalhoun committed:
feat: Only retry media uploads if a network connection is available

Retrying media uploads without a network connection will undoubtedly fail again.

Co-authored-by: Antonis Lilis <antonis.lilis@automattic.com>

derekblank committed:
Merge branch 'trunk' into feat/automatically-retry-pending-media-uploads

derekblank reviewed (approved):
I tested this feature successfully on an internal build, and automated retries are working. Also confirmed that the feature is not active on non-debug/internal builds. 🚀

I removed the RELEASE-NOTES entry, keeping with the precedent that we will add public-facing release notes when the DEBUG flag is removed for this feature.
